backend/data_adapter_test.go

```diff
@@ -0,0 +1,74 @@
+package backend
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
+	"github.com/grafana/grafana-plugin-sdk-go/genproto/pluginv2"
+	"github.com/stretchr/testify/require"
+)
+
+type fakeDataHandlerWithOAuth struct {
+	cli *http.Client
+	svr *httptest.Server
+}
+
+func newFakeDataHandlerWithOAuth() *fakeDataHandlerWithOAuth {
+	settings := DataSourceInstanceSettings{}
+	opts, err := settings.HTTPClientOptions()
+	if err != nil {
+		panic("http client options: " + err.Error())
+	}
+	cli, err := httpclient.New(opts)
+	if err != nil {
+		panic("httpclient new: " + err.Error())
+	}
+
+	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Header.Get("Authorization") == "" {
```

Comment:
If this for whatever reason would not be called you would not get a failing test. Would suggest refactor the test.

Maybe you can use similar test we do have in core Grafana?
https://github.com/grafana/grafana/blob/7f6a1c06a62ea68de3bccfde0ad7f1c32bd5207e/pkg/services/pluginsintegration/clientmiddleware/httpclient_middleware_test.go#L205-L221

Reply:
One nit about that test, when doing:

```
209				reqClone := req.Clone(req.Context())
```

That is already cloning the headers in the req, so the call in 210 is only reinjecting the headers:

```
210				res, err := middlewares[0].CreateMiddleware(httpclient.Options{}, finalRoundTripper).RoundTrip(reqClone)
```

Meaning that the test will success regardless if the `RoundTrip` gets executed. To really verify that the headers are injected at that point, the `Header` from the clone should be cleaned up before.

I will do that in the test here :)

Reply:
I'm not sure about that. The request being cloned is only this without any headers set https://github.com/grafana/grafana/blob/7f6a1c06a62ea68de3bccfde0ad7f1c32bd5207e/pkg/services/pluginsintegration/clientmiddleware/httpclient_middleware_test.go#L169

Feel free to change this as you see if it's actually an error by me :+1: